File: Assets/Lua/colliderHelper.lua
ColliderHelper = ColliderHelper or {}
--@region
local Helper = CS.ColliderHelper
--@endregion

function ColliderHelper:GetHelper(GameObject)
    return Helper.Get(GameObject)
end

--- OnCollisionEnter,OnCollisionStay,OnCollisionExit
function ColliderHelper:BindCollider(helper, ...)
    if helper == nil then return end
    local args = {...}
    helper.luaCollisionEnter = args[1]
    helper.luaCollisionStay = args[2]
    helper.luaCollisionExit = args[3]
end

--- OnTriggerEnter,OnTriggerStay,OnTriggerExit
function ColliderHelper:BindTrigger(helper, ...)
    if helper == nil then return end
    local args = {...}
    helper.luaTriggerEnter = args[1]
    helper.luaTriggerStay = args[2]
    helper.luaTriggerExit = args[3]
end

File: Assets/Scripts/ColliderHelper.cs
using System;
using UnityEngine;


public class ColliderHelper : MonoBehaviour
{
    public Action<Collision> luaCollisionEnter, luaCollisionStay, luaCollisionExit;
    public Action<Collider> luaTriggerEnter, luaTriggerStay, luaTriggerExit;

    public static ColliderHelper Get(GameObject go)
    {
        ColliderHelper helper = go.GetComponent<ColliderHelper>() == null ? 
            go.AddComponent<ColliderHelper>():go.GetComponent<ColliderHelper>();
        return helper;
    }

    private void OnCollisionEnter(Collision collision) => luaCollisionEnter?.Invoke(collision);

    private void OnCollisionStay(Collision collision) => luaCollisionStay?.Invoke(collision);

    private void OnCollisionExit(Collision collision) => luaCollisionExit?.Invoke(collision);

    private void OnTriggerEnter(Collider other) => luaTriggerEnter?.Invoke(other);

    private void OnTriggerStay(Collider other) => luaTriggerStay?.Invoke(other);

    private void OnTriggerExit(Collider other) => luaTriggerExit?.Invoke(other);
}
